modules/pam_userdb/pam_userdb.c

```diff
@@ -129,6 +137,76 @@ _pam_parse (pam_handle_t *pamh, int argc, const char **argv,
   return ctrl;
 }
 
+/*
+ * Database abstraction functions
+ */
+static void *
+db_open(const char *database, mode_t file_mode)
+{
+    void *dbm;
+
+#ifdef HAVE_GDBM_H
+    dbm = gdbm_open(database, 4096, GDBM_READER, file_mode, NULL);
+#else
+    dbm = dbm_open(database, O_RDONLY, file_mode);
+#endif /* HAVE_GDBM_H */
+
+    return dbm;
+}
+
+static datum
+db_firstkey(void *dbm)
+{
+    datum key;
+
+#ifdef HAVE_GDBM_H
+    key = gdbm_firstkey(dbm);
+#else
+    key = dbm_firstkey(dbm);
+#endif /* HAVE_GDBM_H */
+
+    return key;
+}
```

Comment:
In all these new wrapper functions, do they really need that level of complexity?
For example, `db_firstkey` could be simplified to

```
static datum
db_firstkey(void *dbm)
{
#ifdef HAVE_GDBM_H
    return gdbm_firstkey(dbm);
#else
    return dbm_firstkey(dbm);
#endif
}
```

Reply:
Updated